### ISSUE TITLE
Speed on GPX treadmill run is too fast and does not match GPX[BUG]

### ISSUE THREAD

**Issue by SOLLY9** [cagnulein/qdomyos-zwift#1870]
The GPX run option does not use the correct speed. It was far too fast setting off at sub 4 minute miles highest my machine would go!

I attach the debug log for this run, i cant download the GPX file I used for the run but enclose a snapshot of QZ tiles early on in the run which I ended and also the run from garmin connect which I used

My treadmill is nordictrack 2950 22 speed max model.

[debug-Thu_Dec_7_18_32_53_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/13603809/debug-Thu_Dec_7_18_32_53_2023.log)

![Screenshot_20231207-183326](https://github.com/cagnulein/qdomyos-zwift/assets/150450430/41560351-6fb7-4f74-95e9-2b1429aaef24)
![Screenshot_20231207-185221_Connect](https://github.com/cagnulein/qdomyos-zwift/assets/150450430/3f31964c-280a-4f95-9158-46867b74bd3f)
![Screenshot_20231207-185235_Connect](https://github.com/cagnulein/qdomyos-zwift/assets/150450430/5ea6efe4-594f-4f7d-9c6f-831c6079b173)

**Comment by cagnulein** [comment]
hi @SOLLY9 i need the gpx to check the issue, otherwise I can't check it. Thanks!

**Comment by SOLLY9** [comment]
> hi @SOLLY9 i need the gpx to check the issue, otherwise I can't check it. Thanks!

Hi I'm not sure how to get it to you. I tried but gpx is unsupported on github. Do you have a solution?

**Comment by SOLLY9** [comment]
I compressed the file😊
[COURSE_240684197.zip](https://github.com/cagnulein/qdomyos-zwift/files/13604117/COURSE_240684197.zip)

**Comment by SOLLY9** [comment]
Ok. Please let me know when it is updated

**Comment by SOLLY9** [comment]
The gpx file was from a run using garmin fenix 6 pro sapphire watch and garmin HRM pro strap. The data recording setting would have been the watch default setting  "SMART" whatever that is. It's worth noting that the data recording has two options. The other being ",EVERY SECOND". I guess the default helps conserve battery for longer runs or rides. There may be other settings also for when ultra events are selected?

**Comment by cagnulein** [comment]
yes the one second one is the standard one, give it a try.
in the meantime i will check the gpx that you sent if it's fixable somehow
Let me know

Il giorno ven 8 dic 2023 alle 09:22 SOLLY9 ***@***.***> ha
scritto:

**Comment by SOLLY9** [comment]
Ok I will do my next outdoor run at second intervals. If QZ cannot be made to calculate differently based on which time option has been selected it might be worth confirming this on the manual to stop further queries by others. Let me know the outcome also👍

**Comment by cagnulein** [comment]
nobody reads the manual :D it's better to handle it somehow

Il giorno ven 8 dic 2023 alle 09:44 SOLLY9 ***@***.***> ha
scritto:

**Comment by yellowbarchetta** [comment]
Agree with @cagnulein there's a problem with the raw file here.  If I import that GPX into a running website it says that the first mile has been covered at 2m53 per mile.  How does that original gpx file run look on whatever website or service you usually log it on (e.g. Strava, Garmin Connect)?

**Comment by SOLLY9** [comment]
Its from Garmin Connect. The data is in these pictures
![Screenshot_20231208-111823_Connect](https://github.com/cagnulein/qdomyos-zwift/assets/150450430/38900a5f-7f02-4870-9340-ac762f536e01)
![Screenshot_20231208-111835_Connect](https://github.com/cagnulein/qdomyos-zwift/assets/150450430/75ce97b5-d000-41d7-b5f2-aa2b5ebcb44a)
![Screenshot_20231208-111848_Connect](https://github.com/cagnulein/qdomyos-zwift/assets/150450430/b23f79fa-d4ea-4a88-95a7-9f00d3e6de0b)
![Screenshot_20231208-111903_Connect](https://github.com/cagnulein/qdomyos-zwift/assets/150450430/b1904729-c774-4c6a-8a93-4cce91502080)
![Screenshot_20231208-111912_Connect](https://github.com/cagnulein/qdomyos-zwift/assets/150450430/070ca2ed-35b6-497d-9adf-5f6f6ac28a56)
![Screenshot_20231208-111942_Connect](https://github.com/cagnulein/qdomyos-zwift/assets/150450430/d2b1e0ac-3750-4ebb-b0aa-d322ff30048d)
![Screenshot_20231208-111955_Connect](https://github.com/cagnulein/qdomyos-zwift/assets/150450430/17dcc526-cd98-4011-8339-04d8429395b9)

**Comment by yellowbarchetta** [comment]
That's wierd - from GC on the website can you export the .fit file (that's the raw file that the watch has uploaded) - because the GPX is definitely very odd.  e.g. the first 40 data points are all in the first second of the run (18:11:43 - that might not be local time) and yet there is distance calculated between those, hence the unrealistic speed at the start.  
But Garmin Connect has processed it differently.  I know that GC uses the "speed" recorded in the file rather than calculating the distance between point a and point b and dividing that by the time taken to compute speed, whereas other sites like Strava do the second method.  The benefit of Garmin's method is that it's laps will show correctly.

**Comment by yellowbarchetta** [comment]
Have a look at this spreadsheet - I've converted the GPX into a spreadsheet and calculated the distances between each point recorded, then summarised it by second.
https://docs.google.com/spreadsheets/d/1Em40zWRuLrugSCJmI5EVJIlzHJt5iFy9Qg48N3J4-dw/edit?usp=sharing
You can see that in the first second of the run it has recorded you running 190m, then 57m, then 74m, the 62m etc.  Insane speeds!  I can't explain why.  But later in the run it gets down to sensible distances per second (3m per second is about 9min/mile).  btw, nice location to run - my son's raced at Witton Park in the past in XC races (plenty of mud!) and I'm originally from Bury, though down in leafy Cheshire these days.  
Bottom line is I think that that single gpx file is borked for some reason, so it's not a QZ issue.

**Comment by SOLLY9** [comment]
This file is "export original" file. I guess its fit but it exported zipped
[12072689353.zip](https://github.com/cagnulein/qdomyos-zwift/files/13613544/12072689353.zip)

**Comment by SOLLY9** [comment]
I run witton way a lot.😊

This file might not work but I tried others and they were also too fast. Originally I tried a fit file but it was confirmed QZ doesn't use them so I tried a few GPX ones.

**Comment by yellowbarchetta** [comment]
That fit file looks fine.  So it's just the GPX which looks dodgy.  I've imported both into a website I use a lot (FetchEveryone) which allows two traces to be played side by side and the GPX one tears off into the distance, whereas the fit file steadily tracks down the same route.  How did you get the gpx?  Maybe there's something odd in the way that you export them which is causing dodgy data to be written to the gpx file?

**Comment by SOLLY9** [comment]
I just used the export box file in connect.
![Screenshot_20231208-115442_Samsung Internet](https://github.com/cagnulein/qdomyos-zwift/assets/150450430/2bb603bc-1ffc-4b85-b8bb-b67d7ae93ceb)

**Comment by yellowbarchetta** [comment]
That's very odd then.  More-so in that it gave a different date for the activity in the GPX file than it did in the FIT file.  The GPX says it was 7th December, but the fit file was 23rd September?

**Comment by SOLLY9** [comment]
I did the run in September. 7th December was yesterday when I probrably sent the gpx to you which I zipped then. Might be the zip date?

**Comment by SOLLY9** [comment]
[activity_12072689353.zip](https://github.com/cagnulein/qdomyos-zwift/files/13614686/activity_12072689353.zip)

**Comment by SOLLY9** [comment]
I just sent you the gpx file zipped directly from the run. Does that one look correct?

**Comment by yellowbarchetta** [comment]
Yes, that "ACTIVITY" one looks correct - the time stamps go up by individual seconds, the time stamps reflect the September activity etc.

It's the "COURSE" one which was messed up by Garmin - presumably they are only intending the "COURSE" exports to be routes, rather than timed / paced activities.  

Here's a side by side of my COURSE and ACTIVITY - you can see that the locations are the same, but the times are wildly different, and in the COURSE one it goes up in milliseconds.
![image](https://github.com/cagnulein/qdomyos-zwift/assets/58736713/1b224a3e-02f3-4790-8e6e-0f7e58ddb014)

**Comment by SOLLY9** [comment]
I think you are right. The treadmill is a bit faster on the activity gpx than the actual run pace was but it's much closer. 

I will be trail running tomorrow so will check that run after and if it's ok close this down.

Might have to put this one down to user error🤦‍♂️

**Comment by yellowbarchetta** [comment]
You can use the speed offset in the QZ settings to slow it down if the treadmill doesn't match up well, btw.  I do this (in reverse) on my treadmill as it needs to run faster to be closer to the real world.

Glad we found the issue though!

**Comment by cagnulein** [comment]
thanks @yellowbarchetta for the complete analysis!

**Comment by SOLLY9** [comment]
Hi again. I tried a few runs from activity and unfortunately they are still running too fast maybe up to 40 to 50% higher. 

These were done straight from activity in garmin and even though I tried to offset the speed in the app it made no difference. I can't use the gpx runs unless i can get the speed to match or be close to the original runs.

**Comment by SOLLY9** [comment]
> You can use the speed offset in the QZ settings to slow it down if the treadmill doesn't match up well, btw. I do this (in reverse) on my treadmill as it needs to run faster to be closer to the real world.
> 
> Glad we found the issue though!

Where are the speed settings in QZ? I tried various settings in advanced settings both speed offset and speed gain but they did nothing for the treadmill speed.

**Comment by cagnulein** [comment]
> Hi again. I tried a few runs from activity and unfortunately they are still running too fast maybe up to 40 to 50% higher.
> 
> These were done straight from activity in garmin and even though I tried to offset the speed in the app it made no difference. I can't use the gpx runs unless i can get the speed to match or be close to the original runs.

please post always a debug log and the gpx file otherwise it's impossible for me to help you. Thanks @SOLLY9

**Comment by cagnulein** [comment]
> Where are the speed settings in QZ? I tried various settings in advanced settings both speed offset and speed gain but they did nothing for the treadmill speed.

for example if you set speed gain to 0.8, you are reducing the speed read from the treadmill by the 20%

**Comment by SOLLY9** [comment]
Yes it reduces what the speed shows in QZ but it doesn't reduce the treadmill speed unfortunately

**Comment by cagnulein** [comment]
hah ok yes for this you have to use the minus in the target speed tile. details here https://github.com/cagnulein/qdomyos-zwift/issues/1091

**Comment by yellowbarchetta** [comment]
Happy to take a look at the GPX and see if there is anything odd about it if you upload to here @SOLLY9

**Comment by SOLLY9** [comment]
I just went back on the treadmill to try the speed target tile. Unfortunately QZ will no longer upload the GPX file to the treadmill. Before it used to confirm loaded correctly and then show the tiles. Now it says nothing and shows the tiles and the gpx is not there. Not sure if you did an update on Beta that is causing this?

**Comment by cagnulein** [comment]
@SOLLY9 when you have an issue please upload always the debug log, otherwise it's impossible to me to understand the issue that you're facing. Thanks

**Comment by SOLLY9** [comment]
[debug-Thu_Dec_21_12_08_34_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/13741844/debug-Thu_Dec_21_12_08_34_2023.log)
[debug-Thu_Dec_21_12_09_19_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/13741845/debug-Thu_Dec_21_12_09_19_2023.log)
[debug-Thu_Dec_21_12_10_21_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/13741846/debug-Thu_Dec_21_12_10_21_2023.log)
[debug-Thu_Dec_21_12_12_13_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/13741847/debug-Thu_Dec_21_12_12_13_2023.log)
[debug-Thu_Dec_21_12_13_51_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/13741848/debug-Thu_Dec_21_12_13_51_2023.log)
[debug-Thu_Dec_21_12_17_42_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/13741849/debug-Thu_Dec_21_12_17_42_2023.log)
[debug-Thu_Dec_21_12_19_41_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/13741851/debug-Thu_Dec_21_12_19_41_2023.log)
[debug-Thu_Dec_21_12_20_23_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/13741852/debug-Thu_Dec_21_12_20_23_2023.log)
[debug-Thu_Dec_21_12_21_22_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/13741854/debug-Thu_Dec_21_12_21_22_2023.log)
[debug-Thu_Dec_21_12_22_11_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/13741855/debug-Thu_Dec_21_12_22_11_2023.log)
[debug-Thu_Dec_21_12_31_24_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/13741856/debug-Thu_Dec_21_12_31_24_2023.log)

These are all today's debug files.

**Comment by cagnulein** [comment]
@SOLLY9 are you on android 14?

```
Thu Dec 21 12:09:50 2023 1703160590994 Debug: qrc:/GPXList.qml onAccepted You chose: content://com.cxinventor.file.explorer.fileprovider/root/storage/emulated/0/Runs/Base 50mins.gpx
Thu Dec 21 12:09:50 2023 1703160590995 Debug: ../src/homeform.cpp static QString homeform::getFileNameFromContentUri(const QString &) getFileNameFromContentUri "content://com.cxinventor.file.explorer.fileprovider/root/storage/emulated/0/Runs/Base 50mins.gpx"
Thu Dec 21 12:09:51 2023 1703160591001 Debug: ../src/homeform.cpp static QString homeform::copyAndroidContentsURI(QUrl, QString) "Base 50mins.gpx" "" ""
Thu Dec 21 12:09:51 2023 1703160591006 Debug: ../src/homeform.cpp static QString homeform::copyAndroidContentsURI(QUrl, QString) copy "/storage/emulated/0/Android/data/org.cagnulen.qdomyoszwift/files/gpx/" false true false
Thu Dec 21 12:09:55 2023 1703160595259 Debug: qrc:/GPXList.qml onClicked folder is /storage/emulated/0/Android/data/org.cagnulen.qdomyoszwift/files/gpx
Thu Dec 21 12:10:12 2023 1703160612003 Debug: qrc:/GPXList.qml onAccepted You chose: 
Thu Dec 21 12:10:12 2023 1703160612008 Debug: ../src/homeform.cpp static QString homeform::copyAndroidContentsURI(QUrl, QString) copy "/storage/emulated/0/Android/data/org.cagnulen.qdomyoszwift/files/gpx/" false false false
```

if so don't put the file in the gpx directory of qz, otherwise qz can't open it. Place for example on the document folder and click on the other folder button on qz and open it from there

Let me know

**Comment by SOLLY9** [comment]
I am on android 9

**Comment by SOLLY9** [comment]
It's in a folder I created in main storage called runs. It loaded last week

**Comment by SOLLY9** [comment]
I did the run at 70% speed target. Probrably a bit slow. The inclines were out of sync with the run. I wonder if its because the run was slowed and the inclines happened st yhr original gpx times. I attach the treadmill run, the original gpx and the debug file which may show what happens to the incline timing.


[activity_13179949339.zip](https://github.com/cagnulein/qdomyos-zwift/files/13744485/activity_13179949339.zip)

[Base 50mins.zip](https://github.com/cagnulein/qdomyos-zwift/files/13744483/Base.50mins.zip)

[debug.zip](https://github.com/cagnulein/qdomyos-zwift/files/13744533/debug.zip)

**Comment by cagnulein** [comment]
the incline has an average of 100m so maybe you are seeing this?
anyway the file loading worked moving in the qz folder, right?
i will travel all the day long tomorrow, so i will check this in the next
days
thanks

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 21 dic 2023 alle 18:15 SOLLY9 ***@***.***> ha
scritto:

**Comment by SOLLY9** [comment]
Thanks for the help. 

Yes speed can now be adjusted😊

**Comment by cagnulein** [comment]
@SOLLY9 just to understand: how did you understand that the incline is not in sync? from the map? it could be that it 100meters ahead?

**Comment by SOLLY9** [comment]
Hi again. The speeds were out for the inclines I ran. They were too fast and didnt match correctly to the map. It may be that its 100m ahead causing it.

**Comment by cagnulein** [comment]
if you want i can create a version without the average or i can also
average the speed. what do you think? the downside of don't have the
average is that the treadmill will change a lot during the workout

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mar 26 dic 2023 alle 13:56 SOLLY9 ***@***.***> ha
scritto:

**Comment by SOLLY9** [comment]
It might be worth averaging the speed. Could you reduce the 100 metres at all?

**Comment by cagnulein** [comment]
I can do whatever we want :)
If you want I can create a version without the average so you can check if
that's the issue

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mer 27 dic 2023 alle 13:22 SOLLY9 ***@***.***> ha
scritto:

**Comment by SOLLY9** [comment]
I would be happy to test it to see if that fixes it and let you know

**Comment by cagnulein** [comment]
@SOLLY9 done https://github.com/cagnulein/qdomyos-zwift/actions/runs/7338437980


Google is very slow in these days, reviewing the app on the store, even the beta ones. For this reason, in order to check right now if the patch that I did is right, you can use the android build in the link. In order to test it you need first to remove the current QZ version on your phone, install the one that you downloaded.
This version is a 15 minutes trial (you can restart it as many times as you want) and I can unlock it if you want, send me an email to roberto.viola83@gmail.com mentioning this ticket and that you are on android.
Let me know if you have any questions

**Comment by cagnulein** [comment]
@SOLLY9 any news?

**Comment by SOLLY9** [comment]
Hi there. I haven't done much due to hurting my arm. I tried to download the link today but could not find a patch just a kind of workflow diagram so I cant try it so far.

**Comment by cagnulein** [comment]
sorry to hear it! i hope you will feel better soon! this is the apk https://github.com/cagnulein/qdomyos-zwift/actions/runs/7338437980/artifacts/1136413483

**Comment by SOLLY9** [comment]
Hi. I downloaded it but will need to configure it to my system. I'm not sure 15 minutes lifespan for doing that and then trialling it is viable.

**Comment by cagnulein** [comment]
you can restart the app after 15 minutes. you can do this all the time.
Anyway I can unlock it if you want, just send me an email to
roberto.viola83 at gmail.com

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 11 gen 2024 alle ore 16:22 SOLLY9 ***@***.***>
ha scritto:

**Comment by cagnulein** [comment]
@SOLLY9 Did you try this?
